content/blog/copilot-lessons/index.md

```diff
@@ -0,0 +1,131 @@
+
+**Debugging Deployments:** One of the first questions our internal users asked was, 'Why did my latest infrastructure deployment fail?' LLMs clearly excel at summarization, and extracting a clear natural language explanation from stack traces and logs is a clear win.
+
+**Understanding Complex Infrastructure:** Copilot helped our engineers gain insights into our own infrastructure. Asking, 'How many resources are in production?' 'What expensive compute is running' or 'What version are the EKS clusters in EU?" shows the value of allowing users to express infrastructure questions in natural language.
+
+**Generating Code:** One of the first queries logged was, 'I want a static website on AWS behind a CloudFront CDN.' Another was a Solutions Engineers, tasked with demonstrating Pulumi's CrossGuard policy engine to a prospect, asking Copilot to generate policy code.
+
+These early experiences showed the value of CoPilot. But they also revealed the need for a systematic approach to handling diverse user queries. This led us to develop what we call skills.
+
+## Skillful Slicing: Modular Mastery
+
+As Copilot grew, we broke it into smaller pieces we call skills. Each skill does one specific job. The Insights skill handles queries about resource usage and configuration ("How many S3 buckets do I have?"), the Cloud Skill interacts with the Pulumi Service API to manage infrastructure ("Create a new project."), the Code Skill generates Pulumi code snippets ("Write a Typescript program..."), and the Docs Skill retrieves information from Pulumi documentation ("How do I use [update plans](https://www.pulumi.com/docs/iac/concepts/update-plans/)?").  
+
+When you ask Copilot something, it figures out what you need and picks the right skill for the job – like a manager deciding which expert to send your question to. This function-calling approach, orchestrated by a component we call the "outer loop," allows Copilot to access and process information beyond its internal knowledge base.
+
+To train this routing system, we analyzed thousands of user queries and mapped them to Pulumi's underlying functionality. A question like 'Show me my untagged EC2 instances' breaks down into clear parameters - resource type (EC2), filter condition (untagged) - that route to the Insights skill. This mapping helped us handle the many ways users phrase the same technical request.
+
+Refining this routing system revealed another opportunity: streamlining the Debug button workflow.
+
+## Debug Dispatch
+
+![Before and After](optimize.png)
+
+Originally, when a user clicks "Debug with Copilot" on a failed update, we would send a text query to CoPilot like "Analyze this update and explain any errors." The LLM then:
+
+1. Determines the user wants to analyze an update
+2. Identifies which API to call
+3. Calls the API
```

Comment:
Technically it does not call the API, we do based on step 2 then feed the results back to the LLM.